GAE/src/org/waterforpeople/mapping/app/web/DataProcessorRestServlet.java

```diff
@@ -137,8 +139,28 @@ protected RestResponse handleRequest(RestRequest req) throws Exception {
             QuestionGroup newQuestionGroup = qgDao.getByKey(dpReq.getQuestionGroupId());
             QuestionGroup originalQuestionGroup = qgDao.getByKey(Long.valueOf(dpReq.getSource()));
             if (originalQuestionGroup != null && newQuestionGroup != null) {
+                
+                //To avoid collisions, collect all ids already in use in this survey group
+                Set<String> idsInUse = new HashSet<>();
+                final SurveyDAO sDao = new SurveyDAO();
+                final QuestionDao qDao = new QuestionDao();
+                
+                Long surveyId = originalQuestionGroup.getSurveyId();
+                Survey s0 = sDao.getById(surveyId);
+                final Long surveyGroupId = s0.getSurveyGroupId();
+                List<Survey>sList = sDao.listSurveysByGroup(surveyGroupId);
+                for (Survey s : sList) {
+                    List<QuestionGroup>qgList = qgDao.listQuestionGroupBySurvey(s.getKey().getId());
+                    for (QuestionGroup qg : qgList) {
+                        List<Question> qList = qDao.listQuestionsInOrderForGroup(qg.getKey().getId());
```

Comment:
Given the ordering of questions is not necessarily needed for this particular step, I suggest we skip querying for the question groups and directly retrieve the list of questions for a particular survey (form) with `QuestionDao().listQuestionsBySurvey()`